python/kvikio/_lib/libnvcomp.pyx

```diff
@@ -57,305 +63,184 @@ class pyNvcompType_t(Enum):
-        temp_arr,
-        temp_bytes,
-        out_arr,
-        out_bytes,
-        uintptr_t stream=0
+        decomp_buffer,
+        comp_buffer,
     ):
-        cdef uintptr_t in_ptr = Array(in_arr).ptr
-        cdef uintptr_t temp_ptr = Array(temp_arr).ptr
-        cdef uintptr_t out_ptr = Array(out_arr).ptr
-        self.d.decompress_async(
-            <void*>in_ptr,
-            <size_t>in_bytes,
-            <void*>temp_ptr,
-            <size_t>temp_bytes,
-            <void*>out_ptr,
-            <size_t>out_bytes,
-            <cudaStream_t>stream)
-
+        self._impl.decompress(
+            <uint8_t*><uintptr_t>decomp_buffer.data.ptr,
+            <const uint8_t*><uintptr_t>comp_buffer.data.ptr,
+            <DecompressionConfig&>self._decompression_config.get()[0]
+        )
 
-# LZ4 Compressor / Decompressor
-cdef class _LZ4Compressor:
-    cdef __LZ4Compressor* c
+    def set_scratch_buffer(self, new_scratch_buffer):
+        return self._impl.set_scratch_buffer(
+            <uint8_t*><uintptr_t>new_scratch_buffer.data.ptr
+        )
 
-    def __cinit__(self, size_t chunk_size=0):
-        self.c = new __LZ4Compressor()
+    def get_required_scratch_buffer_size(self):
+        return self._impl.get_required_scratch_buffer_size()
 
-    def __dealloc__(self):
-        del self.c
+    def get_compressed_output_size(self, comp_buffer):
+        return self._impl.get_compressed_output_size(
+            <uint8_t*><uintptr_t>comp_buffer.data.ptr
+        )
 
-    def configure(self, in_bytes, temp_bytes, out_bytes):
-        cdef uintptr_t temp_bytes_ptr = Array(temp_bytes).ptr
-        cdef uintptr_t out_bytes_ptr = Array(out_bytes).ptr
-        self.c.configure(
-            <size_t>in_bytes,
-            <size_t*>temp_bytes_ptr,
-            <size_t*>out_bytes_ptr)
 
-    def compress_async(
+cdef class _ANSManager(_nvcompManager):
+    def __cinit__(
         self,
-        in_arr,
-        in_bytes,
-        temp_arr,
-        temp_bytes,
-        out_arr,
-        out_bytes,
-        uintptr_t stream=0
+        size_t uncomp_chunk_size,
+        user_stream,
+        const int device_id,
     ):
-        cdef uintptr_t in_ptr = Array(in_arr).ptr
-        cdef uintptr_t temp_ptr = Array(temp_arr).ptr
-        cdef uintptr_t out_ptr = Array(out_arr).ptr
-        cdef uintptr_t out_bytes_ptr = Array(out_bytes).ptr
-        self.c.compress_async(
-            <void*>in_ptr,
-            <size_t>in_bytes,
-            <void*>temp_ptr,
-            <size_t>temp_bytes,
-            <void*>out_ptr,
-            <size_t*>out_bytes_ptr,
-            <cudaStream_t>stream)
-
-cdef class _LZ4Decompressor:
-    cdef __LZ4Decompressor* d
-
-    def __cinit__(self):
-        self.d = new __LZ4Decompressor()
-
-    def __dealloc__(self):
-        del self.d
+        self._impl = <nvcompManagerBase*>new ANSManager(
+            uncomp_chunk_size,
+            <cudaStream_t><void*>0,  # TODO
+            device_id
+        )
```

Comment:
Would suggest raising for unsupported `user_stream` values as opposed to ignoring them. This way when they are implemented the error can be dropped and the rest of the code kept as-is

Also this can be applied in other compressors below

```suggestion
        if user_stream != 0:
            raise NotImplementedError(
                f"Only `user_stream` of `0` is supported currently. Got: {user_stream}"
            )
        self._impl = <nvcompManagerBase*>new ANSManager(
            uncomp_chunk_size,
            <cudaStream_t>user_stream,
            device_id
        )
```